nsis/src/nsis/file/NsisExecutable.java

```diff
@@ -35,49 +42,117 @@ public NsisExecutable() {
 			byte[] content = reader.readByteArray(headerOffset,
-					NsisConstants.NSIS_MAGIC.length);
-			if (Arrays.equals(NsisConstants.NSIS_MAGIC, content)) {
-				return headerOffset;
+					NsisConstants.NSIS_SIGINFO.length + NsisConstants.NSIS_MAGIC.length);
+			if (Arrays.equals(Bytes.concat(NsisConstants.NSIS_SIGINFO, NsisConstants.NSIS_MAGIC),
+					content)) {
+				return headerOffset - StructConverter.DWORD.getLength(); // Include
+																			// flags
+																			// in
+																			// header
 			}
 		}
 		throw new InvalidFormatException("Nsis magic not found.");
 	}
 
-	private void initScriptHeader(ByteProvider bp)
-			throws IOException, InvalidFormatException {
-		BinaryReader br = new BinaryReader(bp, /* isLittleEndian= */ true);
-		br.setPointerIndex(this.headerOffset);
-		this.scriptHeader = new NsisScriptHeader(br);
+	private void initScriptHeader() throws IOException, InvalidFormatException {
+		this.reader.setPointerIndex(this.headerOffset);
+		this.scriptHeader = new NsisScriptHeader(this.reader);
+	}
+
+	/**
+	 * Attempt to decompress the data from the reader. Supports LZMA algorithm. Will
+	 * eventually support Bzip2 and Zlib.
+	 * 
+	 * @throws IOException
+	 */
+	private void decompressData() throws IOException {
+		byte compressionByte = this.reader.readNextByte();
+		long compressedDataOffset;
+		int compressedDataLength;
+		byte[] compressedData;
+		if (isLZMA(compressionByte)) {
+			int dictionarySize = this.reader.readNextInt();
+			compressedDataOffset = this.headerOffset + NsisScriptHeader.getHeaderSize()
+					+ NsisConstants.COMPRESSION_LZMA_HEADER_LENGTH;
+			compressedDataLength = (this.scriptHeader.compressedHeaderSize ^ 0x80000000)
+					- NsisConstants.COMPRESSION_LZMA_HEADER_LENGTH; // Flip the
+																	// MSB to
+																	// get the
+																	// length
+			compressedData = this.reader.readByteArray(compressedDataOffset, compressedDataLength);
+			decompressLZMA(compressedData, compressionByte, dictionarySize);
+		} else if (isBzip2(compressionByte)) {
+			// TODO Bzip2 decompress
+			System.out.println("Decompress Bzip");
+		} else {
+			// TODO Zlib decompress
+			System.out.println("Decompress Zlib");
+		}
+		return;
+	}
+
+	private boolean isLZMA(byte significantByte) {
+		return NsisConstants.COMPRESSION_LZMA == significantByte;
+	}
+
+	private boolean isBzip2(byte significantByte) {
+		return NsisConstants.COMPRESSION_BZIP2 == significantByte;
+	}
+
+	/**
+	 * Decompressed LZMA bytes using a known properties byte and dictionary size.
+	 * The properties byte is the first byte in the LZMA header and the dictionary
+	 * size corresponds to the 4 following bytes.
+	 * 
+	 * @param compressedData
+	 * @param propByte       the byte indicating LZMA properties
+	 * @param dictionarySize the size of the dictionary to use for decompression
+	 * @throws IOException
+	 */
+	private void decompressLZMA(byte[] compressedData, byte propByte, int dictionarySize)
```

Comment:
prefer to return decompressed data instead of implicitly setting the `decompressedBytes` field. In this case, at the call site:
this.decompressedBytes = decompressLZMA(....)

This will also make testing easier.